### DISCUSSION TITLE
Group Rewrite Proposal

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#7299] [Ideas]
I was thinking of a next generation for `Group` in light of issues and features I need working in nested objects (e.g. #7130 #7142).
I wanted to understand the considerations that lead to designing `Group` as it is. In particular, why does `Group` transform objects to group's coordinate plane and not a different approach?
I have in mind an object, let's call it `DrawingGroup | VirtualGroup`, that will extend `Collection` and it's functionality will differ from `Group` in one major aspect. It will pre-transform the rendering context so that all objects will remain in the canvas coordinate plane. It will need `Group` resizing logic as well as a mechanism to apply transformations to it's objects.
I think this will fix lots of issues regarding geometry and more for nested objects and will unlock powerful features with ease (e.g. #6776) and can eventually deprecate `Group`.
@asturur what do you think?
I saw #6316 that has an interesting solution as well.

**Comment by asturur** [comment]
fabric.Group was here before i joined, so pre 2014.
Probably it was designed so that it could work.
There was no caching, objects had to be drawn all each render so having the objects recentered around the group was a way to do so.

Group has virtually no features apart add and remove.
So if we want to try to create a better container with a different name and then eventually swap or rename, that is totally doable.

I understand that having the objects changing coordinates when grouped is boring and is an annoyance.

**Comment by ShaMan123** [comment]
I tried #7316. IMO it can't serve as a viable solution. There are too many cases that need ugly workarounds.

**Comment by asturur** [comment]
What are the cases? the Layer can complement the Group.
The group can loose some functionality and the Layer can have some but not all.

**Reply by ShaMan123** [reply]
It's hard to keep transform, position and opacity of objects in sync when layer changes.
I listened to the `modified` event but it doesn't cover imperative changes to layer.
Positioning logic is difficult and confusing (it's always hard for me).
It is very awkward.
I tried to replace a major object I use in production which extends group (group is a bad fit for the object). I dumped it after a couple of hours. 

The main pains with group imposes are:
1. Layout strategy which can be mitigated by overriding `_calcBounds` (I believe a PR can make this more straight forward)
2. object geometry - intersections etc (can be addressed by exposing methods on group or fixing geometry logic).
3. `absolutePositioned`
4. object controls + selectability

The thing is group is heavily embedded into fabric. Everywhere. So replacing isn't easy.

**Reply by ShaMan123** [reply]
It is possible, I need help...

**Reply by asturur** [reply]
Ok let's start in a more organized way.
I expect the following:
- a dev will create a Layer doing: `new fabric.Layer([....objects], options)`
- no transformations of objects happen to objects when they are added to layer. They are just referenced, they are still in the stack of the canvas can can be individually selected. ( right/wrong? )

Let's start from fabric.StaticCanvas and let's forget till the staticCanvas part is done, what should happen with interaction.
- i can assign, top, left, scaleX, scaleY, angle, skewX, skewY to Layer ( right/wrong?)
- i can change opacity of Layer ( right/wrong ?)

Are those requirements a very basic version of your idea?
If yes i can start with the questions.

**Reply by ShaMan123** [reply]
I agree with all except:
> they are still in the stack of the canvas

The objects should be in the layer stack IMO.
However they should be individually selected and the layer should be selectable as well in some manner of UX

**Reply by asturur** [reply]
ok so let's forget selecting for now and let's build a prototype of this Layer that render objects as a group but without touching top and left and acoords.
That would be the first step, even if it does not work.
Let's make it in a branch, so we can merge quickly

**Reply by ShaMan123** [reply]
#7316

**Reply by ShaMan123** [reply]
I'll add that I think it should render as Rect does, with options like fill and shadow

**Reply by asturur** [reply]
That wouldn't be a layer. You can set shadow to all objects.

**Comment by ShaMan123** [comment]
#7449
